docs/architecture/adr-080-libp2p-implementation.md

```diff
@@ -0,0 +1,157 @@
+# ADR 80: Libp2p Implementation Plan
+
+## Changelog
+
+- 2022-05-20: Initial Draft.
+
+## Status
+
+Accepted
+
+## Context
+
+Adopting libp2p is an larger project that is the culmination of a long
+programme of work in the tendermint codebase.
+
+## Alternative Approaches
+
+None. 
+
+The decision to adopt libp2p was part of ADR 73, and this ADR lays out
+an implementation plan and records a collection of decisions related
+to this implementation which have not been formally collected in any
+other form. 
+
+## Decisions
+
+### Feature Parity
+
+The goal of this project should be a p2p stack that has feature parity
+with the legacy stack. Everything that is possible with the p2p stack
+today should be possible in the future, eventually reducing the
+overhead of peer communication and complexity in the gossip/peer layer
+of the system. 
+
+In persuit of this goal, this project has two explicit *non-goals*:
+
+- Do not support message/wire compatibility with legacy stack.
+- Do not support dual-mode (legacy/libp2p) in a single node or network.
+
+There does not seem to be a compelling use case of having a release
+that contains both P2P stacks. While either of these goals could
+be theoretically possible, by eliminating both of these goals we can
+effectively reduce the scope of the project and increase our
+confidence.
+
+### Target Iterative Progress 
+
+Rather than produce an extensive design document (other than this
+record of decisions,) we will work iteratively in a topic 
+branch--`main/libp2p`--to prototype the new libp2p integration. This
+work will aim to: 
+
+- develop a working (by way of the e2e tests), implementation of the
+  p2p layer without requiring changes to the reactor implementations.
+- migrate the reactor tests after completing a prototype. 
+- address design questions as they come up, by amending this document
+  or in periodic conversations over patches. 
+- minimize or avoid any interface changes between the reactor
+  implementations and the peer layer. 
+
+### Use libp2p Idiomatically 
+
+
+
+### Minimize P2P Code
+
+One of the largest benefits of this project is that we should be able
+to delete the vast majority of the existing p2p code, which should
+reduce the maintenance burden within Tendermint.
+
+### Defer Transport Decisions
+
+There are many high level aspects of p2p systems such as handshakes between
+nodes, transport protocols, which are fully configurable in the
+context of libp2p. There is no need to make decisions about these at
+this time, our decisions. 
+
+We should not, however, (necessarily) aim to delegate these decisions
+to the users. While we should make it possible for users to pass a
+custom libp2p connection object (`host.Host`) with limited support, we
+should **not** aim to offer too many customizations or configurations
+of different libp2p fundamentals.
```

Comment:
I'd suggest saying this even more strongly, e.g.,

```suggestion
We will not make these decisions user-configurable without a specific need.
We will make it possible for users to pass a custom libp2p connection object
(`host.Host`) with limited support, but will otherwise not expose configurations
and customizations of different libp2p fundamentals.
```

Reply:
Down the road, it would be useful to know exactly what customisations we will explicitly disallow and what the reason for this is. I think it will help us also understand whether any of the libp2p features introduces potentialy vulnerabilities for tendermint.